pyo3-ffi/src/cpython/lock.rs

```diff
@@ -0,0 +1,10 @@
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct PyMutex {
+    pub _bits: u8,
+}
```

Comment:
It'd be wrong to derive Copy and/or Clone, the [documentation](https://docs.python.org/3.14/c-api/init.html#c.PyMutex) explicitly says it should not be moved.   

The Debug impl + not using atomics means the debug impl would race.

The field is also not public in the C api, I'd be inclined to keep it that way here as well.

```suggestion
#[repr(C)]
#[derive(Default)]
pub struct PyMutex {
    _bits: u8,
    _pin: PhantomPinned,
}
```

Reply:
Unfortunately that won't work because `PyObject` needs those traits. At least if we want to say that `ob_mutex` is a `PyObject` as I have it in this version of the PR. Maybe better to write the `PyObject` bindings with `ob_mutex` as a `u8`. It's an internal, private CPython field and we wouldn't be using it otherwise.

Reply:
I think we should probably remove `Copy` & `Clone` from `PyObject` in that case, those traits sound to me like an accident waiting to happen anyway 👍

Reply:
Agreed

Reply:
The `Default` implementation seems a bit risky to me because of the existence of `std::mem::take`. I'm not sure `_pin` will save us from that unless we don't give users a way to get at `&mut PyMutex`, which e.g. they could just do with:

```rust
let mut m = PyMutex::default();
let moved = std::mem::take(&mut mutex);
```

This would also be a move-related bug:

```rust
let mut m = PyMutex::default();
PyMutex_Lock(&mut m);
let moved = m;
```

If we want to allow users to construct a PyMutex, I think we need to remove the `Default` implementation and have something like

```rust
impl PyMutex {
    pub fn new() -> Pin<Box<PyMutex>>
}
```

... but I also wonder whether that's overkill and we should just 

1. not bother with `_pin`, and
2. not give users a way to construct `PyMutex` for now (we will need it inside the crate for `PyObject_HEAD_INIT`, but we can just make `pub(crate) _bits`).

Reply:
>  I'm not sure _pin will save us

The `PhantomPinned` itself doesn't do anything. It lets us build safe abstractions on top of it.

> If we want to allow users to construct a PyMutex

We can remove the `Default` impl and document that users can create it with `mem::zeroed`, which is what CPython's documentation does.

> This would also be a move-related bug:

A safe api would probably do something like this (to follow the example of having a PyMutex on the stack), which is possible because of the `PhantomPinned`.

```rust
let mutex: Pin<&mut PyMutex> = pin!(PyMutex {
     _bits: 0,
     _pin: PhantomPinned
 });
    
 lock(mutex.as_ref());
```

Reply:
> Unfortunately that won't work because PyObject needs those traits

@ngoldbaum we can still have a Debug impl, it just needs a manual impl that correctly accesses the PyMutex value (or, ideally it should just be an AtomicU8). 

I agree with removing the Copy and Clone impls.